// @geist-ui/react/dist/use-toasts/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function (sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread2(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = null != arguments[i] ? arguments[i] : {};
    i % 2 ? ownKeys(Object(source), !0).forEach(function (key) {
      _defineProperty(target, key, source[key]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) {
      Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
    });
  }

  return target;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _slicedToArray(arr, i) {
  return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();
}

function _toConsumableArray(arr) {
  return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread();
}

function _arrayWithoutHoles(arr) {
  if (Array.isArray(arr)) return _arrayLikeToArray(arr);
}

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

function _iterableToArray(iter) {
  if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter);
}

function _iterableToArrayLimit(arr, i) {
  var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"];

  if (_i == null) return;
  var _arr = [];
  var _n = true;
  var _d = false;

  var _s, _e;

  try {
    for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];

  return arr2;
}

function _nonIterableSpread() {
  throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}

var useCurrentState = function useCurrentState(initialState) {
  var _useState = React.useState(function () {
    return typeof initialState === 'function' ? initialState() : initialState;
  }),
      _useState2 = _slicedToArray(_useState, 2),
      state = _useState2[0],
      setState = _useState2[1];

  var ref = React.useRef(initialState);
  React.useEffect(function () {
    ref.current = state;
  }, [state]);

  var setValue = function setValue(val) {
    var result = typeof val === 'function' ? val(ref.current) : val;
    ref.current = result;
    setState(result);
  };

  return [state, setValue, ref];
};

var defaultParams = {
  toasts: [],
  toastHovering: false,
  updateToasts: function updateToasts(t) {
    return t;
  },
  updateToastHoverStatus: function updateToastHoverStatus() {}
};
var GeistUIContent = /*#__PURE__*/React__default["default"].createContext(defaultParams);
var useGeistUIContext = function useGeistUIContext() {
  return React__default["default"].useContext(GeistUIContent);
};

var getId = function getId() {
  return Math.random().toString(32).slice(2, 10);
};

var defaultToast = {
  delay: 2000
};
var destoryStack = [];
var maxDestoryTime = 0;
var destoryTimer;

var useToasts = function useToasts() {
  var _useGeistUIContext = useGeistUIContext(),
      updateToasts = _useGeistUIContext.updateToasts,
      toastHovering = _useGeistUIContext.toastHovering,
      toasts = _useGeistUIContext.toasts;

  var _useCurrentState = useCurrentState(toastHovering),
      _useCurrentState2 = _slicedToArray(_useCurrentState, 3),
      setHovering = _useCurrentState2[1],
      hoveringRef = _useCurrentState2[2];

  React.useEffect(function () {
    return setHovering(toastHovering);
  }, [toastHovering]);

  var destoryAll = function destoryAll(delay, time) {
    /* istanbul ignore next */
    if (time <= maxDestoryTime) return;
    clearTimeout(destoryTimer);
    maxDestoryTime = time;
    destoryTimer = window.setTimeout(function () {
      /* istanbul ignore next */
      updateToasts(function (currentToasts) {
        if (destoryStack.length < currentToasts.length) {
          return currentToasts;
        }

        destoryStack = [];
        return [];
      });
      clearTimeout(destoryTimer);
    }, delay + 350);
  };

  var setToast = function setToast(toast) {
    var id = "toast-".concat(getId());
    var delay = toast.delay || defaultToast.delay;

    var cancel = function cancel(id, delay) {
      updateToasts(function (currentToasts) {
        return currentToasts.map(function (item) {
          if (item.id !== id) return item;
          return _objectSpread2(_objectSpread2({}, item), {}, {
            willBeDestroy: true
          });
        });
      });
      destoryStack.push(id);
      destoryAll(delay, performance.now());
    };

    updateToasts(function (currentToasts) {
      var newToast = _objectSpread2(_objectSpread2({}, toast), {}, {
        id: id,
        delay: delay,
        cancel: function (_cancel) {
          function cancel() {
            return _cancel.apply(this, arguments);
          }

          cancel.toString = function () {
            return _cancel.toString();
          };

          return cancel;
        }(function () {
          return cancel(id, delay);
        })
      });

      return [].concat(_toConsumableArray(currentToasts), [newToast]);
    });

    var hideToast = function hideToast(id, delay) {
      var hideTimer = window.setTimeout(function () {
        if (hoveringRef.current) {
          hideToast(id, delay);
          return clearTimeout(hideTimer);
        }

        cancel(id, delay);
        clearTimeout(hideTimer);
      }, delay);
    };

    hideToast(id, delay);
  };

  return [toasts, setToast];
};

exports["default"] = useToasts;
